feat: automated ejector

0x0aa0 opened the pull request:
Here we introduce the EjectionManager contract, the purpose of which is to enforce a ratelimit on stake for ejections called automatically by an offchain component. The contract prioritizes successful execution over the race conditions of changes in total stake and regular operator deregistration.

0x0aa0 committed:
feat: ejector

0x0aa0 committed:
improvments

- trailing window
- try/catch ejection

0x0aa0 committed:
percentage

0x0aa0 committed:
fix

gpsanant commented on src/Ejector.sol:
```diff
@@ -0,0 +1,149 @@
+ * @title Used for automated ejection of operators from the RegistryCoordinator
+ * @author Layr Labs, Inc.
+ */
+contract Ejector is IEjector, Ownable{
```
should be upgradeable

gpsanant commented on src/interfaces/IEjector.sol:
```diff
@@ -0,0 +1,61 @@
+ * @title Interface for a contract that ejects operators from an AVSs RegistryCoordinator
+ * @author Layr Labs, Inc.
+ */
+interface IEjector {
```
EjectionManager might be better to differentiate from the "ejector" role

gpsanant commented on src/interfaces/IEjector.sol:
```diff
@@ -0,0 +1,61 @@
+    /// @notice A quorum's ratelimit parameters
+    struct QuorumEjectionParams {
+        uint32 timeDelta; // Time delta to track ejection over
+        uint16 ejectableStakePercent; // Max stake to be ejectable per time delta
```
`rateLimitWindow`?

gpsanant commented on src/Ejector.sol:
```diff
@@ -0,0 +1,149 @@
+                    require(canEject(operatorStake, quorumNumber), "Ejector: Stake exceeds quorum ejection ratelimit");
+                }
+
+                stakeEjectedForQuorum[quorumNumber].push(StakeEjection({
```
we can for sure make this 1 sstore?

gpsanant commented on src/Ejector.sol:
```diff
@@ -0,0 +1,149 @@
+     * @param _operatorIds The addresses of the operators to eject
+     * @param _quorumBitmaps The quorum bitmaps for each respective operator
+     */
+    function ejectOperators(bytes32[] memory _operatorIds, uint256[] memory _quorumBitmaps) external {
```
i prefer something like
`uint256 quorumBitmap, bytes32[quorumNumbers][numOperators] operatorIds`
then we sum up the stake of operators for each quorum and boot them out if low enough and time delta has passed.

seems much simpler and just need to store the previous ejection time

0x0aa0 committed:
nits

0x0aa0 committed:
Merge branch 'm2-mainnet' into auto-eject

0x0aa0 committed:
Merge branch 'dev' into auto-eject

0x0aa0 committed:
feat: non signing metric helpers

- adds operatorId on reg and dereg events
- adds function to OperatorStateRetriever to get bitmaps for a set of operators at a timestamp

0x0aa0 committed:
fix: conflict

0x0aa0 committed:
Merge branch 'dev' into auto-eject

0x0aa0 committed:
refactor: ejection

gpsanant commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,170 @@
+     * @dev This function will eject as many operators as possible without reverting
+     * @param _operatorIds The ids of the operators to eject for each quorum
+     */
+    function ejectOperators(bytes32[][] memory _operatorIds) external {
```
i feel like `ejectOperators(uint8 quorumNumber, bytes32[] calldata _operatorIds)` is more in line with our style

0x0aa0 replied:
https://github.com/Layr-Labs/eigenlayer-middleware/pull/146#discussion_r1456445550

gpsanant replied:
you caught me red handed. now i realize why i wanted that is we want to eject from multiple quorums at the same time

gpsanant replied:
we could do that outside of the contract too, but ok

gpsanant commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,170 @@
+                    break;
+                }
+                
+                try registryCoordinator.ejectOperator(
```
why would we not revert if this failed? why would this fail?

0x0aa0 replied:
Would fail if the operator leaves the set before ejection is called

gpsanant replied:
ah gpgp

0x0aa0 committed:
fix: stake recording

jianoaix commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,163 @@
+
+    /**
+     * @notice Ejects operators from the AVSs RegistryCoordinator under a ratelimit
+     * @param _operatorIds The ids of the operators to eject for each quorum
```
What's assumed in this parameter by this function?
From the impl below it looks: 1) the caller provides all operators eligible for ejection, but delegate to this function for rate limiting; 2) the operators in each quorum are ordered by priority to eject (in case not all can be ejected, it'll just eject a leading prefix of the array)
Might be good to document the assumptions so the caller knows how to pass the params.

0x0aa0 committed:
Merge branch 'dev' into auto-eject

0x0aa0 committed:
test: unit

0x0aa0 committed:
style: natspec

ChaoticWalrus commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,168 @@
+                //if caller is ejector enforce ratelimit
+                if(
+                    msg.sender == ejector && 
+                    quorumEjectionParams[quorumNumber].rateLimitWindow > 0 &&
```
seems like this means that no limit will be enforced for new quorums until a limit is explicitly set on this contract for the quorum?
possibly a bit of a "footgun"?
please correct me if I'm reading this wrong though.

0x0aa0 replied:
That is correct. My thinking here is that when adding a new quorum, while it is finding its footing we will not want to run automated ejection but reserve the ability to eject without ratelimit in that period. Once the set is more stable the quorum params can be set and offchain can be turned on to run regularly.

ChaoticWalrus commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,168 @@
+                --i;
+            }
+        }
+        return totalEjectable - totalEjected;   
```
is there a chance of (reverting on) underflow here, e.g. if `ejectableStakePercent` is edited?
seems fine but might be a rather opaque failure?

0x0aa0 replied:
good call

ChaoticWalrus reviewed (approved):
LGTM overall. Noted a couple more minor edge cases / quirks.
I reviewed the tests briefly but not in detail.

gpsanant commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,168 @@
+        uint256 i;
+        if (stakeEjectedForQuorum[_quorumNumber].length == 0) {
+            return totalEjectable;
+        } else {
```
nit: take out of else

gpsanant commented on src/EjectionManager.sol:
```diff
@@ -0,0 +1,168 @@
+                    quorumEjectionParams[quorumNumber].rateLimitWindow > 0 &&
+                    stakeForEjection + operatorStake > amountEjectable
+                ){
+                    stakeEjectedForQuorum[quorumNumber].push(StakeEjection({
```
why are we pushing and not continuing here?

0x0aa0 replied:
Because we do not want to revert if the stake of the ops in the input is greater than the ratelimit. Here we checkpoint the stake ejected inside the ratelimit and then end execution

gpsanant replied:
oh my bad, i completely misread this

0x0aa0 committed:
fix: totalEjected > totalEjectable

nit: else
